File: config/config_parser/src/config_parser/id.go
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// This file contains the code that computes ids from names.
//
// We use the Fowler-Noll-Vo hash function

package config_parser

import (
	"hash/fnv"
)

func IdFromName(name string) uint32 {
	hash := fnv.New32()
	hash.Write([]byte(name))
	return hash.Sum32()
}

File: config/id.cc
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "config/id.h"

#include <string>

namespace cobalt {
namespace config {

const uint32_t kFnvPrime = 0x1000193;
const uint32_t kFnvOffsetBasis = 0x811c9dc5;

uint32_t IdFromName(const std::string &name) {
  uint32_t hash = kFnvOffsetBasis;
  for (size_t i = 0; i < name.size(); i++) {
    hash *= kFnvPrime;
    hash ^= name[i];
  }
  return hash;
}

}  // namespace config
}  // namespace cobalt
